### PR TITLE
add lb uuid to docluster spec

### LINKED ISSUE #290: Store LB uuid in DOCluster spec
The DO LB uuid is currently stored in status. Upon applying a manifest with the specification to use an existing LB (like restore), CAPDO creates a new LB and tries to update the control plane endpoint. Allow for specifying an existing LB in DOCluster by storing the lb uuid in the spec.  

Slack: https://kubernetes.slack.com/archives/CTND9FHU2/p1639075525031700

### PR DESCRIPTION
**Pull request by varshavaradarajan**
**What this PR does / why we need it**:

DO LB uuid is currently stored in status. Upon restore, we lose this information and capdo creates a new LB instead of re-using the existing LB. This PR adds the lb uuid in the spec. And updates the docluster controller to prefer the spec uuid over status if present. 

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #290 

**Special notes for your reviewer**:
- I only made this change for v1beta1. And ran `make generate`, got a few warnings like:
```
E1228 13:28:36.045121   30566 conversion.go:755] Warning: could not find nor generate a final Conversion function for sigs.k8s.io/cluster-api-provider-digitalocean/api/v1beta1.DOLoadBalancer -> github.com/kubernetes-sigs/cluster-api-provider-digitalocean/api/v1alpha4.DOLoadBalancer
```

- I am yet to test this. The setup we use internally in DO still uses v1alpha4, which makes it difficult for me to test this. 

**Documentation**:

**Release note**:

```
Allow specifying an existing DO LB in docluster spec
```

### CODE REVIEW

File: api/v1beta1/types.go
Comment (MorrisLaw) [inline api/v1beta1/types.go:131]: I _think_ we may need to add this field to the struct in `v1alpha4/types.go` too. It's likely what's causing the conversion warning you're seeing when running `make generate`

**Review by MorrisLaw — COMMENTED** [review]
Thanks for the issue and subsequent PR 🎉

File: api/v1beta1/types.go
Comment (MorrisLaw) [inline api/v1beta1/types.go:131]: Please rename this to `ResourceID`

**Review by MorrisLaw — CHANGES_REQUESTED** [review]

File: api/v1beta1/types.go
Comment (MorrisLaw) [reply api/v1beta1/types.go:131]: Possibly `v1alpha3/types.go` too

**Review by timoreimann — COMMENTED** [review]
Do we need to do something extra to ensure that the new spec field is only ever populated once?

File: api/v1beta1/types.go
Comment (MorrisLaw) [reply api/v1beta1/types.go:131]: > What's our policy on updating APIs as old as v1alpha3? 

I'm not 100% sure. Maybe you're right that worrying about v1alpha3 is unnecessary. @cpanato or @prksu might know for sure? Either way it'd be good to have an explicit rule regarding that.

File: controllers/docluster_controller.go
Comment (varshavaradarajan) [reply controllers/docluster_controller.go:138]: Yes, I left this intentionally so that we don't break older api objects. Once we can be sure that we're no longer relying on the status, we can get rid of it.

File: api/v1beta1/types.go
Comment (varshavaradarajan) [reply api/v1beta1/types.go:131]: @cpanato need your 👀 ^^

**Comment by cpanato** [comment]
@MorrisLaw @MorrisLaw @varshavaradarajan  this PR will be good to have for the 1.1.0 release

**Comment by cpanato** [comment]
@varshavaradarajan please run `make generate` and push the updated files

File: api/v1beta1/types.go
Comment (cpanato) [reply api/v1beta1/types.go]: for olders api, i think we don't need to do that, otherwise might break users, but in the conversion we should add that

@fabriziopandini it is the correct approach when adding a new field in the newer API and not in the older ones? and just write some conversions? or what is the best practice for that? thanks in advance

**Comment by cpanato** [comment]
also will be good to run some manual tests to cover this case, @varshavaradarajan can you write down the steps needed to reproduce this use case?

**Comment by cpanato** [comment]
> @cpanato - I have been running `make generate` but doesn't change any files. :(

ok, I will check this tomorrow my morning :)

**Comment by varshavaradarajan** [comment]
@cpanato - to reproduce:

1. Setup a management cluster x
2. Setup a workload cluster
3. Do clusterctl backup with `clusterctl -v 10 backup -n <workload_cluster_namespace> --directory /tmp/test-backup` with the kubeconfig pointing to the management cluster x.
4. Setup another management cluster y
5. Make management cluster x inaccessible by deleting the apiserver, power off management cluster x's control plane node etc. Do not delete x. The workload cluster should still be intact.
6. Restore  using clusterctl restore: `clusterctl -v 10 restore --directory /tmp/test-backup` . kubeconfig should point to management cluster y.

Expected Result:
Should re-use the existing control plane LB, not create a new one. 

clusterctl restore had a bug which was fixed in v1.0.1: https://github.com/kubernetes-sigs/cluster-api/releases/tag/v1.0.1. Please build a binary with the fix.

**Comment by varshavaradarajan** [comment]
@cpanato - generated the conversion files through `make generate` finally! Waiting for your approval.

**Comment by cpanato** [comment]
> @cpanato - generated the conversion files through `make generate` finally! Waiting for your approval.

thanks for your patience, I would like to run some upgrade tests and this case as well, so bear with me :)

**Comment by cpanato** [comment]
@varshavaradarajan can you please rebase this PR with the latest main branch? we made some fixes: https://github.com/kubernetes-sigs/cluster-api-provider-digitalocean/pull/299 
after that I will finish the manual tests and upgrade tests

thanks for your patience

**Comment by cpanato** [comment]
> 6\. clusterctl -v 10 restore --directory /tmp/test-backup

trying the backup but getting

```
$ clusterctl backup --directory /tmp/test-backup -n default -v 10
Using configuration File="/Users/cpanato/.cluster-api/clusterctl.yaml"
Performing backup...
Discovering Cluster API objects
Secret Count=8
ConfigMap Count=1
Total objects Count=12
Excluding secret from move (not linked with any Cluster) name="default-token-ph9ck"
Object won't be moved because it's not included in GVK considered for move kind="KubeadmConfig" name="test-control-plane-r76tb"
Object won't be moved because it's not included in GVK considered for move kind="KubeadmConfig" name="test-md-0-6txhk"
Object won't be moved because it's not included in GVK considered for move kind="KubeadmControlPlane" name="test-control-plane"
Starting backup of Cluster API objects Clusters=0
Moving Cluster API objects ClusterClasses=0
Pausing the source cluster
Pausing the source cluster classes
Saving files to /tmp/test-backup
Resuming the target cluter classes
Resuming the source cluster
Using configuration File="/Users/cpanato/.cluster-api/clusterctl.yaml"
```

**Comment by cpanato** [comment]
the upgrade from v1alpha4 > v1beta1 works!

**Review by cpanato — APPROVED** [review]
thanks for this!!
/approve

@timoreimann and @MorrisLaw you had comments, can you please take another look and see if that looks good for you?

**Comment by timoreimann** [comment]
For posterity, our related Slack discussion is [here](https://kubernetes.slack.com/archives/CTND9FHU2/p1644920088841939).

This CAPI Slack discussion indicates that the backup/restore failure is due to clusterctl, and not our implementation. Approving since all our tests pass.

/lgtm

**Pull request merged**
